erpnext/public/js/utils/serial_no_batch_selector.js

```diff
@@ -0,0 +1,340 @@
+	setup: function() {
+		this.item_code = this.item.item_code;
+		this.qty = this.item.qty;
+		this.make_dialog();
+	},
+
+	make_dialog: function() {
+		var me = this;
+
+		this.data = this.oldest ? this.oldest : [];
+		let title = "";
+		let fields = [
+			{
+				fieldname: 'item_code',
+				read_only: 1,
+				fieldtype:'Link',
+				options: 'Item',
+				label: __('Item Code'),
+				default: me.item_code
+			},
+			{fieldtype:'Column Break'},
+			{
+				fieldname: 'warehouse',
+				fieldtype:'Link',
+				options: 'Warehouse',
+				label: __(me.warehouse_details.type),
+				default: me.warehouse_details.name,
+				onchange: function(e) {
+					me.warehouse_details.name = this.get_value();
+					var batches = this.layout.fields_dict.batches;
+					if(batches) {
+						batches.grid.df.data = [];
+						batches.grid.refresh();
+						batches.grid.add_new_row(null, null, null);
+					}
+				}
+			},
+			{fieldtype:'Column Break'},
+			{
+				fieldname: 'qty',
+				fieldtype:'Float',
+				read_only: 1,
+				label: __(me.has_batch ? 'Total Qty' : 'Qty'),
+				default: me.qty
+			},
+		];
+
+		if(this.has_batch) {
+			title = __("Select Batch Numbers");
+			fields = fields.concat(this.get_batch_fields());
+		} else {
+			title = __("Select Serial Numbers");
+			fields = fields.concat(this.get_serial_no_fields());
+		}
+
+		this.dialog = new frappe.ui.Dialog({
+			title: title,
+			fields: fields
+		});
+
+		this.bind_qty();
+
+		this.dialog.set_primary_action(__('Insert'), function() {
+			me.values = me.dialog.get_values();
+			if(me.validate()) {
+				me.set_items();
+				me.dialog.hide();
+			}
+		});
+
+		this.dialog.show();
+	},
+
+	validate: function() {
+		let values = this.values;
+		if(!values.warehouse) {
+			frappe.throw(__("Please select a warehouse"));
+			return false;
+		}
+		if(this.has_batch) {
+			if(values.batches.length === 0 || !values.batches) {
+				frappe.throw(__("Please select batches for batched item "
+					+ values.item_code));
+				return false;
+			}
+			values.batches.map((batch, i) => {
+				if(!batch.selected_qty || batch.selected_qty === 0 ) {
+					frappe.throw(__("Please select quantity on row " + (i+1)));
+					return false;
+				}
+			});
+			return true;
+
+		} else {
+			let serial_nos = values.serial_no || '';
+			if (!serial_nos || !serial_nos.replace(/\s/g, '').length) {
+				frappe.throw(__("Please enter serial numbers for serialized item "
+					+ values.item_code));
+				return false;
+			}
+			return true;
+		}
+	},
+
+	set_items: function() {
+		var me = this;
+
+		if(this.has_batch) {
+			this.values.batches.map((batch, i) => {
+				let row = (i !== 0) ? this.frm.add_child("items") : this.item;
+				refresh_field("items");
+				if(i !== 0) {
+					// let index = i + me.item.idx - 1;
+					// let grid_row = me.frm.fields_dict.items.grid.grid_rows[index];
+					// grid_row.toggle_editable_row();
+					// grid_row.on_grid_fields_dict.item_code.set_value(this.item_code);
+					row.item_code = this.item_code;
```

Comment:
Discovered that `refresh_field("items")` after this line (setting item_code value) does little to trigger the relevant fields and turn the indicator green.
One way is probably using `set_value()`, but it is too expensive for multiple rows.

![screen shot 2017-06-26 at 3 33 45 pm](https://user-images.githubusercontent.com/5196925/27534583-213973d2-5a85-11e7-955a-221ae70dbd70.png)